utils/mcp_client.py

```
@@ -0,0 +1,677 @@
+    if not resolved_ips or not hasattr(aiohttp_module, "TCPConnector"):
+        return None
+    parsed = urlparse(url)
+    if not parsed.hostname:
+        return None
+    return aiohttp_module.TCPConnector(
+        resolver=_PinnedResolver({parsed.hostname: resolved_ips}),
+        ttl_dns_cache=0,
+    )
+
+
+def _response_peer_ip(response: Any) -> str | None:
```

Ruff (0.15.20)

[warning] Dynamically typed expressions (typing.Any) are disallowed in `response`

(ANN401)